In [25]:
import pandas as pd
import numpy as np
from sklearn.datasets import make_regression

In [26]:
X , y = make_regression(n_samples=1000 , n_features=3 , noise=10 , random_state=42)

In [27]:
print(X.shape)

(1000, 3)


In [28]:
print(y.shape)

(1000,)


In [29]:
X = np.c_[np.ones(X.shape[0]), X]

In [30]:
y = y.reshape(len(y),1)

In [31]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [32]:
def calculate_loss(X, y, weights):
    
    #  Calculate predictions by multiplying inputs (X) with weights:)
    predictions = X.dot(weights)
    
    #  Calculate the errors (difference between predictions and actual values):)
    errors = predictions - y
    
    # Calculate Mean Squared Error (MSE):)
    mse = np.mean(errors ** 2)
    
    #  Divide by 2 to match the gradient calculation format:)
    loss = mse / 2
    
    return loss

Gradient Descent

In [35]:
def gradient_descent(X, y, num_iterations , freq_print):
    alpha = 0.001
    W = np.random.randn(X.shape[1], 1)
    for i in range(num_iterations) :    
        yhat = X.dot(W)
        error = yhat - y
        gradient = X.T.dot(error)
        W = W - alpha * gradient / len(X)
        if i % freq_print == 0:
            print(calculate_loss(X , y , W))
        
        
    return W             

In [36]:
W = gradient_descent(X_test , y_test , 10000 , 1000)

8405.155870757595
1229.9527349777063
230.37348851443667
79.1554192799069
54.417115729699184
50.08227705044065
49.27929951334994
49.12422363920703
49.093377975094285
49.087118741780614


In [37]:
W 

array([[ 1.60778446],
       [98.19392278],
       [81.98373928],
       [25.92519061]])

To test the gradient descent

this function give the most efficient weight and I print the loss for this weight

In [39]:
def least_weight(X, y):
    X_transpose = X.T
    W = np.linalg.inv(X_transpose @ X) @ X_transpose @ y
    
    
    print(calculate_loss(X , y , W))
    

print the output for the test data withe the most efficient weights

In [40]:
least_weight(X_test,y_test)

49.08549457182744


The gradient descent final loss output is almost equal to the least loss output.......................
                                   49.087106110189694  |||  49.08549457182744